cpp/oneapi/dal/table/backend/homogen_table_impl.cpp

```diff
@@ -91,206 +90,417 @@ bool has_array_data_kind(const Policy& policy, const array<Data>& array, const A
+        dst.reset(src, start_pointer, dst_count);
     }
+}
 
-    const auto table_dtype = meta_.get_data_type(0);
-    if (block_dtype == table_dtype && has_array_data_kind(policy, data_, kind)) {
-        if (data_.has_mutable_data()) {
-            auto row_data = reinterpret_cast<Data*>(data_.get_mutable_data());
-            auto row_start_pointer = row_data + rows.start_idx * col_count_;
-            block.reset(data_, row_start_pointer, range_size);
+class block_access_provider {
+private:
+    struct block_info {
+        block_info(int64_t row_count,
+                   int64_t column_count,
+                   int64_t row_offset,
+                   int64_t column_offset)
+                : row_count(row_count),
+                  column_count(column_count),
+                  row_offset(row_offset),
+                  column_offset(column_offset),
+                  element_count(row_count * column_count) {
+            detail::check_mul_overflow(row_count, column_count);
+
+            if (row_count <= 0 || column_count <= 0 || row_offset < 0 || column_offset < 0) {
+                throw dal::range_error("invalid parameters");
+            }
         }
-        else {
-            auto row_data = reinterpret_cast<const Data*>(data_.get_data());
-            auto row_start_pointer = row_data + rows.start_idx * col_count_;
-            block.reset(data_, row_start_pointer, range_size);
+
+        int64_t row_count;
+        int64_t column_count;
+        int64_t row_offset;
+        int64_t column_offset;
+        int64_t element_count;
+    };
+
+    struct origin_info {
+        origin_info(data_type dtype, int64_t row_count, int64_t column_count)
+                : dtype(dtype),
+                  row_count(row_count),
+                  column_count(column_count),
+                  element_count(row_count * column_count) {
+            detail::check_mul_overflow(row_count, column_count);
+
+            if (row_count <= 0 || column_count <= 0) {
+                throw dal::range_error("invalid parameters");
+            }
         }
-    }
-    else {
-        if (block.get_count() < range_size || block.has_mutable_data() == false ||
-            has_array_data_kind(policy, block, kind) == false) {
-            reset_array(policy, block, range_size, kind);
+
+        data_type dtype;
+        int64_t row_count;
+        int64_t column_count;
+        int64_t element_count;
+    };
+
+private:
+    void check_origin_data(const array<byte_t>& origin_data,
+                           int64_t origin_dtype_size,
+                           int64_t block_dtype_size) const {
+        detail::check_mul_overflow(origin_.element_count,
+                                   std::max(origin_dtype_size, block_dtype_size));
+        if (origin_data.get_count() != origin_.element_count * origin_dtype_size) {
+            throw dal::range_error("origin has less data than required");
         }
```

Comment:
Looks like this condition is invariant to the table itself. My impression is that this condition should be validated when the array is passed to construct a homogen table. If you think it's important to test here, put assertion.

PS: Seems condition itself is too restrictive, condition to validate should be `origin_data.get_count() >= origin_.element_count * origin_dtype_size`. In other words, we should allow homogen table creation from the array which element count exceeds `row_count * column_count`